### DISCUSSION TITLE
Separate serialization and validation alias generators

### DISCUSSION THREAD

**Discussion by rndubs** [pydantic/pydantic#7877] [Question] [answered]
Hi Folks. I am attempting to extend `BaseModel` in a way that provides the following characteristics:

1. `snake_case` attribute names (PEP8 compliant)
2. Validation on "lower case with no underscores" aliases that are derived from the attribute name
3. Conversion of input fields to "lower case with no underscores" for validation
4. Serialization aliases derived from the attribute name that use `CamelCase` convention

My main motivation for wanting separate aliases is so that the field names in the schema representation are user-friendly when using tools like [autodoc-pydantic](https://github.com/mansenfranzen/autodoc_pydantic) to document our schema.

Is it possible to achieve this by extending `BaseModel` or leveraging other pydantic features?

I'm using pydantic v2.4.

# Manual Way

The desired validator characteristics can be achieved manually using the `Field` class:

```python
from pydantic import BaseModel, Field

class MyModel(BaseModel):
    field_one: str = Field(
        ...,
        validation_alias="fieldone",
        serialization_alias="FieldOne",
    )
``` 

But this becomes quite verbose for classes with several fields.

The input fields (from dict/JSON) can be converted to "lower case no underscores" using a TypeAdapter, which is demonstrated in the following section.

## Conversion of input fields to lower case

A response from a related issue proposed the following solution for lower-casing fields, which can be tweaked slightly to additionally remove underscores:
- https://github.com/pydantic/pydantic/issues/1147#issuecomment-1722248973

This converts the input fields to lower case without underscores:

```python
from pydantic import TypeAdapter, BaseModel, Field
from typing import Any

class CaseInsensitiveAdapter(TypeAdapter[Any]):
    def validate_python(self, value: Any) -> Any:
        def __lower__(value: Any) -> Any:
            if isinstance(value, dict):
                return {k.lower().replace('_', ''): __lower__(v) for k, v in value.items()}
            return value

        value = __lower__(value)
        return super().validate_python(value)

class MyClass(BaseModel):
    field_one: str = Field(..., validation_alias="fieldone", serialization_alias="FieldOne")

adapter = CaseInsensitiveAdapter(MyClass)
my_obj = adapter.validate_python({"FieldA": "hello"})
print(my_obj)
```

This provides the desired validation behavior as well as the desired serialization alias, but still requires manually specifying separate aliases for each attribute/field.

## Combining with an alias generator

Combining the adapter with an alias generator gets me most of the way there, but doesn't allow for separate serialization and validation aliases:

```python
from pydantic import TypeAdapter, BaseModel, Field, ConfigDict
from typing import Any

def lower_no_underscores(field: str) -> str:
    return field.lower().replace('_', '')

class CaseInsensitiveAdapter(TypeAdapter[Any]):
        model_config = ConfigDict(alias_generator=lower_no_underscores)

    def validate_python(self, value: Any) -> Any:
        def __lower__(value: Any) -> Any:
            if isinstance(value, dict):
                return {lower_no_underscores(k): __lower__(v) for k, v in value.items()}
            return value

        value = __lower__(value)
        return super().validate_python(value)

class MyClass(BaseModel):
    field_one: str

adapter = CaseInsensitiveAdapter(MyClass)
my_obj = adapter.validate_python({"FieldOne": "hello"})
my_obj.model_json_schema()
```

# Options

Are there any other pydantic features I can use to achieve the desired characteristics?

**Comment by cfriedland5** [comment, marked as answer]
I think something like this will do what you want

```python
import inspect

from pydantic import BaseModel, Field
from pydantic.alias_generators import to_camel
from pydantic.fields import FieldInfo


def lower_no_underscores(field: str) -> str:
    return field.lower().replace('_', '')


class DefaultAliasBaseModel(BaseModel):
    """BaseModel that adds default aliases."""

    def __init_subclass__(cls, **kwargs) -> None:
        """Add default aliases to the class."""
        super().__init_subclass__(**kwargs)
        for key in inspect.get_annotations(cls):
            try:
                attr = getattr(cls, key)
            except AttributeError:
                field_info = Field(
                    validation_alias=lower_no_underscores(key), serialization_alias=to_camel(key)
                )
                setattr(cls, key, field_info)
            else:
                if isinstance(attr, FieldInfo):
                    if attr.validation_alias is None:
                        attr.validation_alias = lower_no_underscores(key)
                    if attr.serialization_alias is None:
                        attr.serialization_alias = to_camel(key)
```
